edap.py

```diff
@@ -213,14 +42,181 @@ def __init__(self, hostname, admin_cn, password, domain=None):
         self.SERVICES = "ou=services"
         self.SERVICES_GROUP = f"{self.SERVICES},{self.BASE_DN}"
 
-    def add_s(self, * args, ** kwargs):
-        return self.ldap.add_s(* args, ** kwargs)
-
-    def modify_s(self, * args, ** kwargs):
-        return self.ldap.modify_s(* args, ** kwargs)
-
-    def search_s(self, * args, ** kwargs):
-        return self.ldap.search_s(* args, ** kwargs)
+    def add_s(self, *args, **kwargs):
+        return self.ldap.add_s(*args, **kwargs)
+
+    def modify_s(self, *args, **kwargs):
+        return self.ldap.modify_s(*args, **kwargs)
+
+    def search_s(self, *args, **kwargs):
+        return self.ldap.search_s(*args, **kwargs)
+
+    def unbind_s(self):
+        return self.ldap.unbind_s()
+
+
+class LdapObjectsMixin(object):
+
+    def object_exists(self, search, obj_class=None):
+        if obj_class is not None:
+            search = f"&({search})(objectClass={obj_class})"
+        found = self.ldap.search_s(self.ldap.BASE_DN, ldap.SCOPE_SUBTREE, f"({search})")
+        return len(found)
+
+    def object_exists_at(self, root, obj_class, additional_search=None):
+        search = f"objectClass={obj_class}"
+        if additional_search is not None:
+            search = f"&({search})({additional_search})"
+        try:
+            found = self.ldap.search_s(root, ldap.SCOPE_BASE, f"({search})")
+        except Exception:
+            return 0
+        return len(found)
+
+    def subobject_exists_at(self, relative_pos, obj_class, additional_search=None):
+        root = f"{relative_pos},{self.ldap.BASE_DN}"
+        return self.object_exists_at(root, obj_class, additional_search)
+
+
+class LdapGroupMixin(object):
+
+    def create_org_unit(self, name, fqdn):
+        dic = dict(
+             ou=name.encode("ASCII"),
+             objectclass=(b"organizationalUnit", b"top"),
+        )
+        modlist = ldap.modlist.addModlist(dic)
+        self.ldap.add_s(fqdn, modlist)
+
+    def create_group_dict(self, name):
+        dic = dict(
+            cn=name.encode("ASCII"), objectclass=(b"posixGroup", b"top"), gidNumber=b"500",
+        )
+        return dic
+
+    def create_group_from_dict(self, fqdn, dic):
+        modlist = ldap.modlist.addModlist(dic)
+        self.ldap.add_s(fqdn, modlist)
+
+    def create_service_group(self, name):
+        if not self.subobject_exists_at(self.ldap.SERVICES, "organizationalUnit"):
+            self.create_org_unit(self.ldap.SERVICES, self.ldap.SERVICES_GROUP)
+        if not self.subobject_exists_at(f"cn={name},{self.ldap.SERVICES}", "posixGroup"):
+            dic = self.create_group_dict(f"{name}")
+            self.create_group_from_dict(f"cn={name},{self.ldap.SERVICES_GROUP}", dic)
+
+
+class LdapUserMixin(object):
+
+    def add_user(self, uid, name, surname, password):
+        if self.subobject_exists_at("ou=people", "organizationalUnit") == 0:
```

Comment:
Note for later: The `ou=people` is hardcoded here, but it could be something like a class attribute - something that has default value, but that can be overriden if needed.

Reply:
Thanks, will refactor it in #2 !